### PR TITLE
Add sort to sementic context

### PR DESCRIPTION
**Pull request by aljazerzen**
Introduces sort syntax for `asc` and `desc`:

```elm
from Employees
sort [id]
sort [age, desc:last_name, asc:first_name]
```

Includes refactor in preparation for #300

### CODE REVIEW

**Comment by max-sixty** [comment]
Very nice! Thanks!

**Comment by max-sixty** [comment]
> ```elm
> sort [age, desc:last_name, asc:first_name]
> ```

I'm actually not so sure about this. 

Generally the `lvalue`s are where to put the `rvalue`s — e.g. `e:employees`: "put the `employees` table into the `e` namespace". Or `by:title`: "Put the `title` column in the `by` parameter".

They're not generally _modifiers_ of the `rvalue`s. But here they are. 

That said, I'm not sure how else to do it. We don't currently support (though we could):

```elm
sort [age, last_name order:desc, first_name order:asc]
```

I haven't really seen other languages do this well — either the specific issue of how to specify sort order, or optional boolean arguments / flags. 

Probably the one that most resonates with me for these is shell conventions, which allows an optional flag, e.g. `--desc` (and not `--desc=true`).

**Comment by aljazerzen** [comment]
I don't like this `sort [age, last_name order:desc, first_name order:asc]` because it's hard to parse since it is very similar to a function call of `last_name` with argument `order`.

What about nested arrays:
```
sort [age, [last_name, order:desc], [first_name order:asc]]
sort age
sort [last_name, order:desc]
sort [age, first_name]
```

Or tuples:
```
sort [age, (last_name, order:desc), (first_name order:asc)]
sort age
sort (last_name, order:desc)
sort (age, first_name)
```

**Comment by max-sixty** [comment]
> I don't like this `sort [age, last_name order:desc, first_name order:asc]` because it's hard to parse since it is very similar to a function call of `last_name` with argument `order`.

Good point

Yes, I agree none of those are ideal.

What do you think about `sort [age, last_name:desc, first_name:asc]` over the current approach?

I'm not sure there's enough space in the lang for "flags", but if there were, `sort [age, last_name --desc, first_name --asc]` would work, maybe as sugar for one of the other options you suggest.

**Comment by aljazerzen** [comment]
There is one more option from dyplr:

```
[age, -first_name, last_name]
```

**Comment by max-sixty** [comment]
Great idea! I quite like it; what do you think?

**Comment by aljazerzen** [comment]
It's simple and concise, I am all in for it. 

On the note of what to do with NULLs, dplyr also has some advice:

> ### Missing values
> 
> Unlike base sorting with [sort()](https://rdrr.io/r/base/sort.html), NA are:
> - always sorted to the end for local data, even when wrapped with [desc()](https://dplyr.tidyverse.org/reference/desc.html).
> - treated differently for remote data, depending on the backend.

**Comment by max-sixty** [comment]
Great! Thanks for suggesting the idea!

Re nulls — yeah — how much of that can we do from PRQL? There are some things I'd love to change about nulls, but I think many of them are ingrained in the DB / in SQL standards...

**Comment by aljazerzen** [comment]
Well there is the `ORDER BY col DESC NULLS [FIRST/LAST]`, which we can choose to support or not. 

My suggestion was that for now, we always append `NULLS LAST`
